[[FIX]] Param destructuring should not effect max params

lukeapage opened the pull request:
Fixes destructuring so that multiple destructured params do not count
towards the maxparams limit. This is because destructuring is the solution
to having alot of numbered params, since you would replace a hard to
figure out list of params with an object, then use destructuring to pull
out from that object.
Fixes #2183

nicolo-ribaudo commented:
I think you should also fix `JSHINT.data().functins[x].metrics.parameters `

lukeapage commented:
> I think you should also fix JSHINT.data().functins[x].metrics.parameters

done, good spot.

lukeapage commented:
rebased on master

jugglinmike commented:
The name "count" is a little generic and potentially misleading. Initially, I was wondering why you would track that separately than the `parameters` array itself because I was thinking it was redundant. In reality (and as I'm sure you know), `parameters` describes all bindings from the parameter list, so the number is definitely distinct. What do you think about re-naming `count` to `arity` to more clearly communicate the concept it describes?

lukeapage commented:
> What do you think about re-naming count to arity to more clearly communicate the concept it describes?

yes, arity is a more descriptive name, I'll change

lukeapage commented:
Fixed. Also added another commit - https://github.com/jshint/jshint/commit/f75b2d36f46fa9db913efcf736cbaec3df2b0708
When renaming it occurred to me this was a bit cleaner than polluting the function directly.

jugglinmike commented:
Excellent! Merged to `master` at commit 9d021eede3e5ebffe2c98fce12293703586b491b. Thanks, Luke!
